# A)Data Retrieval and Preprocessing

## Data Retrieval

In [7]:
!pip install kagglehub

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [kagglehub]/2 [kagglehub]


In [ ]:
from pathlib import Path
import pandas as pd
import kagglehub
from kagglehub import KaggleDatasetAdapter

def load_cashflow_data(csv_name: str = "dataset.csv") -> pd.DataFrame:
    # 1) Look for local file first: ../raw_data/dataset.csv
    base_dir = Path.cwd().parent          # one level up from notebook
    raw_data_dir = base_dir / "raw_data"
    raw_data_dir.mkdir(parents=True, exist_ok=True)
    local_path = raw_data_dir / csv_name

    if local_path.is_file():
        print(f"Loading local file: {local_path}")
        return pd.read_csv(local_path)

    # 2) If not found, download from Kaggle once via kagglehub
    print("Local file not found, downloading from Kaggle via kagglehub...")
    df = kagglehub.dataset_load(
        KaggleDatasetAdapter.PANDAS,
        "pradumn203/payment-date-prediction-for-invoices-dataset",
        "dataset.csv",  # main CSV name on Kaggle
    )

    # Save for future offline use
    df.to_csv(local_path, index=False)
    print(f"Saved local copy to {local_path}")

    return df

df = load_cashflow_data()

Local file not found, downloading from Kaggle via kagglehub...
Saved local copy to /home/saiyudh/code/EwaltsJ/cf_copilot/raw_data/dataset.csv


## A.1)Feature Extraction and Engineering

### A.1.1) Quick Look at Data Structure

In [18]:
df.head(2)

,business_code,cust_number,name_customer,clear_date,buisness_year,doc_id,posting_date,document_create_date,document_create_date.1,due_in_date,invoice_currency,document type,posting_id,area_business,total_open_amount,baseline_create_date,cust_payment_terms,invoice_id,isOpen
0,U001,0200769623,WAL-MAR corp,2020-02-11 00:00:00,2020.0,1.930438e+09,2020-01-26,20200125,20200126,20200210.0,USD,RV,1.0,NaN,54273.28,20200126.0,NAH4,1.930438e+09,0
1,U001,0200980828,BEN E,2019-08-08 00:00:00,2019.0,1.929646e+09,2019-07-22,20190722,20190722,20190811.0,USD,RV,1.0,NaN,79656.60,20190722.0,NAD1,1.929646e+09,0


In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   business_code           50000 non-null  object 
 1   cust_number             50000 non-null  object 
 2   name_customer           50000 non-null  object 
 3   clear_date              40000 non-null  object 
 4   buisness_year           50000 non-null  float64
 5   doc_id                  50000 non-null  float64
 6   posting_date            50000 non-null  object 
 7   document_create_date    50000 non-null  int64  
 8   document_create_date.1  50000 non-null  int64  
 9   due_in_date             50000 non-null  float64
 10  invoice_currency        50000 non-null  object 
 11  document type           50000 non-null  object 
 12  posting_id              50000 non-null  float64
 13  area_business           0 non-null      float64
 14  total_open_amount       50000 non-null

Checking values in all the columns

### A.1.2)Peeking into each of the columns

In [26]:
df['business_code'].value_counts()

business_code
U001    45359
CA02     3917
U013      573
U002      135
U005       11
U007        5
Name: count, dtype: int64

In [45]:
print(df['clear_date'].value_counts())
print(df['clear_date'].isnull().value_counts())

clear_date
2019-11-12 00:00:00    309
2019-09-03 00:00:00    249
2019-10-15 00:00:00    246
2019-11-01 00:00:00    242
2019-02-19 00:00:00    241
                      ... 
2019-06-15 00:00:00      1
2020-05-04 00:00:00      1
2019-06-08 00:00:00      1
2020-04-11 00:00:00      1
2019-07-20 00:00:00      1
Name: count, Length: 403, dtype: int64
clear_date
False    40000
True     10000
Name: count, dtype: int64


name_customer -> some names are repeated

clear_date -> False 40000, True 1000

business_data -> only 2019,2020 data

doc_id -> some repition as no missing id

invoice_currency -> USD,CAD

posting_date -> 506 posting dates

## A.2)Data Preprocessing and Wrangling

### A.2.1)Data Cleaning

Checking for duplicates

In [16]:
df.duplicated().value_counts()

False    48839
True      1161
Name: count, dtype: int64

we can see 1161 datas duplicated.

## A.3)Feature Scaling & Selection

# B)ALGO Testing

# C)Model Eval and Tuning

# D)Pipeline